File: Array/3sum-closest.py
from typing import List

# 260 ms 16.76%
# 14.9 32.63%
# 还可以添加剪枝，当sublist左边加起来还比target大，或者右边加起来还比target小 都可以直接break了
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        ans = 0xfffffff
        nums.sort()
        for i in range(len(nums)-2):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            temp = target - nums[i]
            left = i + 1
            right = len(nums) - 1
            while left < right:
                if abs(target - ans) > abs(target - nums[left] - nums[right] - nums[i]):
                    ans = nums[left] + nums[right] + nums[i]
                if nums[left] + nums[right] > temp:
                    right -= 1
                elif nums[left] + nums[right] < temp:
                    left += 1
                else:
                    break
        return ans

if __name__ == '__main__':
    s = Solution()
    nums = [2,2,2,4,7]
    target = 11
    print(s.threeSumClosest(nums, target))

File: Array/3sum-closest.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int ans = 10e5;  // Pay attention to the range
        for (int anchor = 0; anchor < nums.size() - 2; anchor++){
            int left = anchor +1, right = nums.size()-1;
            while (left < right){
                int sum = nums[anchor] + nums[left] + nums[right];
                ans = abs(sum - target) < abs(ans - target) ? sum : ans;
                if (sum < target){
                    left += 1;
                } else{
                    right -= 1;
                }
            }
        }
    return ans;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {1, 1, 1, 1};
    int target = -100;
    int ans = sol.threeSumClosest(nums, target);
    cout << ans << endl;
    return 0;
}